### PR TITLE
Automatic hoisting of common dependencies to repo root

### PR DESCRIPTION
**Pull request by gigabo**
This adds a pair of options to govern installing some or all external dependencies at the repo root instead of distributed in package directories.

From the [original proposal](https://github.com/lerna/lerna/issues/334) for this:

> Hoisting external dependencies that are shared across packages:
> - dramatically speeds up bootstrap
> - reduces the storage footprint of a Lerna repo
> - handles the singleton problem where a common dependency (e.g. React) must be shared

The options may be set on the command lined or in asini.json.

This addresses https://github.com/asini/asini/issues/14.

### CODE REVIEW

File: test/PackageUtilities.js [hunk 1]
```diff
@@ -77,10 +77,18 @@ describe("PackageUtilities", () => {
...
-    it("should throw when --scope is given but empty", () => {
-      assert.throws(() => {
-        PackageUtilities.filterPackages(packages, "");
-      });
+    it("should return everything when --scope is given but empty", () => {
+      assert.deepEqual(
+        PackageUtilities.filterPackages(packages, "").map((pkg) => pkg.name),
+        ["package-3", "package-4", "package-a-1", "package-a-2"]
+      );
+    });
```
Comment (gigabo) [inline test/PackageUtilities.js:85]: This is an change in edge-case behavior due to the way filtration is factored to support hoisting as well as scope/ignore.  Previously `--scope=` was an error.  Now it's a noop.

**Review by roblg — COMMENTED** [review]
LGTM. I'm not super-familiar with the code, but the big things seem to be there, and look reasonable (hoisting/semver matching/warn on differing versions/.bin hangling).

**Comment by gigabo** [comment]
Thanks for looking at this @roblg.  I've since pushed two small fixes for minor issues I discovered while playing around bootstrapping babel, but they're pretty trivial.

The bigger news is hoisting makes bootstrap twice as fast for babel! 🚀

**Comment by roblg** [comment]
> The bigger news is hoisting makes bootstrap twice as fast for babel! 🚀

That's awesome!

**Comment by gigabo** [comment]
Just made it even faster!  Starting the root install _first_ since it's likely to take longest.

Now babel bootstrap takes _12s_ with hoisting compared to _30s_ without on my laptop!

**Comment by gigabo** [comment]
For comparison:  With `yarn` (on an experimental branch I'm working on) it's ~17s without hoisting (slower than the ~12s of `npm` with hoisting).  With `yarn` _plus_ hoisting... _under 7 seconds_. :godmode:

**Comment by gigabo** [comment]
This will be easier to see in ascii art form:

```
              hoist

            no    yes
          +-----+-----+
        no| 30s | 12s |
yarn      |-----+-----+
       yes| 17s |  7s |
          +-----+-----+
```

**Comment by gigabo** [comment]
Been using this with React Server for a couple of days locally.  Do want.

Going to merge this afternoon, barring objections.

File: src/PackageUtilities.js [hunk 2]
```diff
@@ -4,6 +4,7 @@ import Package from "./Package";
 import path from "path";
 import {sync as globSync} from "glob";
 import minimatch from "minimatch";
+import isArray from "isarray";
```
Comment (rygine) [inline src/PackageUtilities.js:7]: why not `util.isArray` (node < 4), or `Array.isArray` (node >= 4)

File: src/commands/BootstrapCommand.js [hunk 3]
```diff
@@ -129,32 +130,251 @@ export default class BootstrapCommand extends Command {
...
+
+          // Get the object for this package, auto-vivifying.
+          const dep = depsToInstall[name] || (depsToInstall[name] = {
+            versions   : {},
+            dependents : {}
+          });
+
+          // Add this version if it's the first time we've seen it.
+          if (!dep.versions[version]) {
+            dep.versions  [version] = 0;
```
Comment (rygine) [inline src/commands/BootstrapCommand.js:237]: this whitespace hurts my brain! 😝

File: src/commands/BootstrapCommand.js [hunk 4]
```diff
@@ -129,32 +130,251 @@ export default class BootstrapCommand extends Command {
...
+        });
+      });
+    }
+
+    // Install anything that needs to go into the leaves.
+    Object.keys(leaves)
+      .map((pkgName) => ({pkg: this.packageGraph.get(pkgName).package, deps: leaves[pkgName]}))
+      .forEach(({pkg, deps}) => actions.push(
+        (cb) => NpmUtilities.installInDir(pkg.location, deps, (err) => {
           this.progressBar.tick(pkg.name);
```
Comment (rygine) [inline src/commands/BootstrapCommand.js:365]: am i missing something, or is this tick not captured in the progress  bar length specified below?

File: src/commands/BootstrapCommand.js [hunk 5]
```diff
@@ -129,32 +130,251 @@ export default class BootstrapCommand extends Command {
...
+    Object.keys(depsToInstall).forEach((name) => {
+      const {versions, dependents} = depsToInstall[name];
+
+      let rootVersion;
+
+      if (scope && PackageUtilities.getFilteredPackage({name}, {scope, ignore})) {
+
+        // Get the most common version.
+        const commonVersion = Object.keys(versions)
+          .reduce((a, b) => versions[a] > versions[b] ? a : b);
```
Comment (rygine) [inline src/commands/BootstrapCommand.js:257]: this is much more elegant than what i did. 👍

File: src/commands/BootstrapCommand.js (hunk 4 above)
Comment (gigabo) [reply src/commands/BootstrapCommand.js:365]: The progress bar is just initialized with the number of actions we're going to run.

``` javascript
this.progressBar.init(actions.length);
```

There's a vestigial comment down there, though.  I'll clean that up.

File: src/PackageUtilities.js (hunk 2 above)
Comment (gigabo) [reply src/PackageUtilities.js:7]: Oh nice, didn't know about `util.isArray`.

This is still easier, though... :D

**Review by rygine — APPROVED** [review]
LGTM

**Comment by gigabo** [comment]
Thanks for looking at this.  Bombs away!
